src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala

```diff
@@ -199,31 +199,34 @@ case class MergeIntoCommand(
     }}.asNondeterministic()
 
     // Skip data based on the merge condition
-    val targetOnlyPredicates =
+    val targetOnlyPredicates: Seq[Expression] =
       splitConjunctivePredicates(condition).filter(_.references.subsetOf(target.outputSet))
     val dataSkippedFiles = deltaTxn.filterFiles(targetOnlyPredicates)
 
-    // Apply inner join to between source and target using the merge condition to find matches
-    // In addition, we attach two columns
-    // - a monotonically increasing row id for target rows to later identify whether the same
-    //     target row is modified by multiple user or not
-    // - the target file name the row is from to later identify the files touched by matched rows
-    val joinToFindTouchedFiles = {
-      val sourceDF = Dataset.ofRows(spark, source)
-      val targetDF = Dataset.ofRows(spark, buildTargetPlanWithFiles(deltaTxn, dataSkippedFiles))
-        .withColumn(ROW_ID_COL, monotonically_increasing_id())
-        .withColumn(FILE_NAME_COL, input_file_name())
-      sourceDF.join(targetDF, new Column(condition), "inner")
-    }
+    // if no target files are skipped, then skip a possibly expensive join and collect
```

Comment:
Even though the join may be expensive, it may still reduce the amount of data collected on the Driver, i.e. this may reduce the scalability of the system. Can we put this behind a feature flag that's off by default?